### PR TITLE
Speed up merlin

### PR DESCRIPTION
**Pull request by rgrinberg**
We're doing the same work for every single module in a library. This PR makes it
do it only once.

### CODE REVIEW

**Comment by rgrinberg** [comment]
@Khady does this PR make a meaningful difference?

**Comment by rgrinberg** [comment]
@voodoos the preprocessing handling in the merlin code seems wrong to me. It seems like we don't handle `per_module` correctly.

File: src/dune_rules/merlin.ml [hunk 1]
```diff
@@ -61,36 +56,40 @@ module Processed = struct
...
-               let sexp = to_sexp config in
-               let open Pp.O in
-               Pp.vbox (Pp.text name ++ Pp.cut ++ Sexp.pp sexp) ++ Pp.newline)
+        let sexp = to_sexp config in
+        let open Pp.O in
+        Pp.vbox (Sexp.pp sexp)
+        ++ Pp.newline
+        ++ Pp.vbox
+             (Pp.concat_map modules ~f:(fun m ->
+                  Pp.text (Module_name.to_string m)))
```
Comment (voodoos) [inline src/dune_rules/merlin.ml:92]: This is use for debugging and testing only, and it feels less useful to have a module's name printed after its configuration (and it brings lot of small changes to the tests). Is there a specific reason why you chose to switch them ?

**Comment by voodoos** [comment]
These changes looks good to me. 

> @voodoos Independently of this pr: the preprocessing handling in the merlin code seems wrong to me. It seems like we don't handle `per_module` correctly.

That's probably true will have look today.

**Comment by Khady** [comment]
@rgrinberg I haven't checked if the merlin files are the same as before, but the the noop warm build went from 7.6s with 2.7.1 to 4.6s with this PR. Part of the gain is from #4064. But I think that 4.6s is at least a whole second faster than the numbers I got for #4064.

File: src/dune_rules/merlin.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/merlin.ml:92]: I just thought it made the tests neater. Now we aren't printing the same configuration over and over.

**Comment by rgrinberg** [comment]
> But I think that 4.6s is at least a whole second faster than the numbers I got for #4064.

Yes, I didn't expect much more than that. A fresh profile with this PR would be helpful. This RP removes a bunch of unnecessary maps and sets operations that overlap with other parts of the code, so the profile should look somewhat different now.

**Comment by Khady** [comment]
I will provide one tomorrow

File: src/dune_rules/merlin.ml (hunk 1 above)
Comment (voodoos) [reply src/dune_rules/merlin.ml:92]: I agree that merging identical configuration is better, I was just complaining about the module names being printed after and not before the configuration.

However I had to re-duplicate for `Per_module` pp configs #4092.

**Comment by rgrinberg** [comment]
Closed in favor of #4092

**Pull request closed without merging**